CHANGELOG.md

```diff
@@ -1,3 +1,14 @@
+## Next
+### Enhancements
+- Added `backoff` feature for reconnecting
+- `send_frame/2` is no longer processed during `opening`, ie. it is queued until `connected` or failure to connect.
```

Comment:
This is actually a breaking change, and potentially a large one in my opinion.

I'm torn on this. I understand that it could make the library more convenient, but it could also cause a message to be sent prematurely before some kind of required connection setup was finished.

Adding another method or keyword option would be my preference but that means more stuff to maintain. It could be useful though. I remain somewhat on the edge.

Reply:
> I'm torn on this. I understand that it could make the library more convenient, but it could also cause a message to be sent prematurely before some kind of required connection setup was finished.

Can you elaborate on this? Maybe I'm missing something but I'm not sure when a message could be sent prematurely because of this?

Initially I had failing tests since they currently assume that when the process is started, the connection is connected and able to send messages. Before I realized that I could make this change with selective receive so that the `send_frame` message waits in the inbox until the connection is connected, I added a feature `WebSockex.await_status(pid, :connected)` so that you could do the following:

```elixir
{:ok, pid} = WebSockex.start_link(url)
WebSockex.await_status(pid, :connected)
WebSockex.send_frame(pid, message)
```
However I removed this and instead relied on selective receive because it means less work for the user and one less change in this PR.

Reply:
`handle_connect` is used to queue up initial messages. By casting to itself, a process can queue up messages to send in a specific order after startup.

By handling `send_frame` after the connection is established, we actually put ourselves at much higher risk for a race condition. A frame could be sent before the initial connection messages were sent.

Reply:
ok, I have 2 thoughts on this.

1. I'm just exploring how this could work better ignoring breaking changes. Maybe first read my comment on `handle_send_result`. If instead of `send_frame` we have `handle_send_result` and `handle_call` which would allow a user to implement their own `send_frame`, and we have `init`, then the user could create a message queue in their state during `init` for startup messages and then `handle_call` could add messages to the queue as well and the queue can be processed in the right order starting from `handle_connect`. I also think that adding `handle_continue` support would be useful here. In fact if we just add `handle_continue` then this can be used from `handle_connect` to process initial messages before `send_frame` messages by calling `:continue` in `handle_connect`. 

2. Without making breaking changes, we need a way to wait for the connection to be established before calling `send_frame`. This can either be done by adding `await_status` as above, or by adding `handle_call` which has access to the current status. The user can theoretically keep track of the current status with `handle_connect` and `handle_disconnect` in their own state, and implement `handle_call` on top of `handle_cast`, but it's quite a lot of extra stuff for each user to add which is already a solved problem in a plain `GenServer`. On the one hand I kinda like adding `await_status` natively because it requires no extra work from the user, but on the other hand I'm really seeing that everything could be handled much better with more flexibility if `WebSockex` is on par with `GenServer`, namely having `init`, `handle_call`, `handle_continue`. 

I was trying to solve all the issues I ran into with adding backoff without adding a lot of other features, but it seems that we'd be better served by first adding the other features that we can then more easily build on here.